mono/mini/iltests.il

```diff
@@ -3308,6 +3308,56 @@ L_3:
 		ret
 	 }
 
+	.method private hidebysig static valuetype [mscorlib]System.Nullable`1<int32> ret_some_vt(int32 val) cil managed
+	 {
+		.maxstack 8
+		ldarg.0
+		newobj instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
+		ldarg.0
+		ldc.i4.0
+		ble.s second_vt
+
+		ldarg.0
+		ldc.i4.s 40
+		add
+		newobj instance void valuetype [mscorlib]System.Nullable`1<int32>::.ctor(!0)
+		pop
+		ret
+
+second_vt_2:
```

Comment:
@lewurm Is this the instruction where we were starting to lose track of vt_sp because of the instruction above ? If that is so and the stack state here is computed based on the stack state following the `ret` above then this is fishy, since the stack state here should be the same as the one at the end of second_vt basic block instead. Does ECMA guarantee that the stack should be empty when starting a basic block after ret ? Otherwise we are still broken.

Reply:
> I I I. 1. 7. 5.    **Backward   branch   constraints** 
> It shall be possible, with a single forward-pass through the CIL instruction stream for any method, to infer the exact state of the evaluation stack at every instruction (where by “state” we mean the number and type of each item on the evaluation stack). 
> In particular, if that single-pass analysis arrives at an instruction, call it location X, that immediately follows an unconditional branch, and where X is not the target of an earlier branch instruction, then the state of the evaluation stack at X, clearly, cannot be derived from existing information. In this case, the CLI demands that the evaluation stack at X be empty.

I would argue that `ret` qualifies as "unconditional branch".